src/commons.jl

```diff
@@ -0,0 +1,199 @@
+	const KEY_WOW64_64KEY = 0x0100
+	const KEY_WOW64_32KEY = 0x0200
+	const KEY_WOW64_RES = 0x0300
+	
+	const KEY_READ = ((SAM.STANDARD_RIGHTS_READ | KEY_QUERY_VALUE | KEY_ENUMERATE_SUB_KEYS | KEY_NOTIFY) & (~SAM.SYNCHRONIZE))
+	const KEY_WRITE = ((SAM.STANDARD_RIGHTS_WRITE | KEY_SET_VALUE | KEY_CREATE_SUB_KEY) & (~SAM.SYNCHRONIZE))
+	const KEY_EXECUTE = ((KEY_READ) & (~SAM.SYNCHRONIZE))
+	const KEY_ALL_ACCESS = ((SAM.STANDARD_RIGHTS_ALL | KEY_QUERY_VALUE | KEY_SET_VALUE | KEY_CREATE_SUB_KEY | KEY_ENUMERATE_SUB_KEYS | KEY_NOTIFY | KEY_CREATE_LINK) & (~SAM.SYNCHRONIZE))
+	const REG_OPTION_RESERVED = 0x00000000
+end
+
+baremodule REG_OPTION
+	const BACKUP_RESTORE = 0x00000004
+	const CREATE_LINK = 0x00000002
+	const NON_VOLATILE = 0x00000000
+	const VOLATILE = 0x00000001
+end
+
+const REG_CREATED_NEW_KEY = 0x00000001
+const REG_OPENED_EXISTING_KEY = 0x00000002
+
+function getroothkey(key::AbstractString)
+	local pos = search(key, '\\')
+	if pos == 0
+		return convert(HKEY, 0)
+	else
+		root_key = uppercase(key[1:(pos-1)])
+	end
+	
+	if root_key == "HKEY_CLASSES_ROOT" || root_key == "HKCR"
+		return ROOT_HKEY.HKEY_CLASSES_ROOT
+	elseif root_key == "HKEY_CURRENT_CONFIG" || root_key == "HKCC"
+		return ROOT_HKEY.HKEY_CURRENT_CONFIG
+	elseif root_key == "HKEY_CURRENT_USER" || root_key == "HKCU"
+		return ROOT_HKEY.HKEY_CURRENT_USER
+	elseif root_key == "HKEY_LOCAL_MACHINE" || root_key == "HKLM"
+		return ROOT_HKEY.HKEY_LOCAL_MACHINE
+	elseif root_key == "HKEY_USERS" || root_key == "HKU"
+		return ROOT_HKEY.HKEY_USERS
+	elseif root_key == "HKEY_PERFORMANCE_DATA"
+		return ROOT_HKEY.HKEY_PERFORMANCE_DATA
+	elseif root_key == "HKEY_PERFORMANCE_TEXT"
+		return ROOT_HKEY.HKEY_PERFORMANCE_TEXT
+	elseif root_key == "HKEY_PERFORMANCE_NLSTEXT"
+		return ROOT_HKEY.HKEY_PERFORMANCE_NLSTEXT
+	elseif root_key == "HKEY_CURRENT_CONFIG" || root_key == "HKCG"
+		return ROOT_HKEY.HKEY_CURRENT_CONFIG
+	elseif root_key == "HKEY_DYN_DATA" || root_key == "HKDD"
+		return ROOT_HKEY.HKEY_DYN_DATA
+	else
+		return convert(HKEY, 0)
+	end
+end
+
+function getsubkey(key::AbstractString)
+	local pos = search(key,'\\')
+	if pos == 0
+		return ""
+	else
+		return key[(pos+1):end]
+	end
+end
+
+function gettype(typ::UInt32)
+	if typ == REG_TYPE.REG_SZ || typ == REG_TYPE.REG_EXPAND_SZ || typ == REG_TYPE.REG_LINK
+		return AbstractString
+	elseif typ == REG_TYPE.REG_MULTI_SZ
+		return Vector{AbstractString}
+	elseif typ == REG_TYPE.REG_DWORD_LITTLE_ENDIAN || typ == REG_TYPE.REG_DWORD_BIG_ENDIAN
+		return UInt32
+	elseif typ == REG_TYPE.REG_QWORD
+		return UInt64
+	elseif typ == REG_TYPE.REG_BINARY
+		return Vector{UInt8}
+	else
+		return 0
+	end
+end
+
+function regopenkeyex(root::HKEY, subkey::AbstractString, sam::REGSAM)
+	local hk = Ref{HKEY}(0)
+	
+	#UNICODE version; xxxExA is ASCII version, but LPCTRSTR = LPCSTR
+	res = ccall((:RegOpenKeyExW, advapi), stdcall,
+	  LONG,
+	  (HKEY, LPCWSTR, DWORD, REGSAM, PHKEY),
+	  root, transcode(Cwchar_t, subkey), 0, sam, hk )
```

Comment:
OK, this is true for input string, but not for output buffers or when I have to pass a null pointer